homeassistant/components/cover/mqtt.py

```diff
@@ -149,103 +148,116 @@ def validate_options(value):
 async def _async_setup_entity(hass, config, async_add_entities,
                               discovery_hash=None):
     """Set up the MQTT Cover."""
-    value_template = config.get(CONF_VALUE_TEMPLATE)
-    if value_template is not None:
-        value_template.hass = hass
-    set_position_template = config.get(CONF_SET_POSITION_TEMPLATE)
-    if set_position_template is not None:
-        set_position_template.hass = hass
-
-    async_add_entities([MqttCover(
-        config.get(CONF_NAME),
-        config.get(CONF_STATE_TOPIC),
-        config.get(CONF_GET_POSITION_TOPIC),
-        config.get(CONF_COMMAND_TOPIC),
-        config.get(CONF_AVAILABILITY_TOPIC),
-        config.get(CONF_TILT_COMMAND_TOPIC),
-        config.get(CONF_TILT_STATUS_TOPIC),
-        config.get(CONF_QOS),
-        config.get(CONF_RETAIN),
-        config.get(CONF_STATE_OPEN),
-        config.get(CONF_STATE_CLOSED),
-        config.get(CONF_POSITION_OPEN),
-        config.get(CONF_POSITION_CLOSED),
-        config.get(CONF_PAYLOAD_OPEN),
-        config.get(CONF_PAYLOAD_CLOSE),
-        config.get(CONF_PAYLOAD_STOP),
-        config.get(CONF_PAYLOAD_AVAILABLE),
-        config.get(CONF_PAYLOAD_NOT_AVAILABLE),
-        config.get(CONF_OPTIMISTIC),
-        value_template,
-        config.get(CONF_TILT_OPEN_POSITION),
-        config.get(CONF_TILT_CLOSED_POSITION),
-        config.get(CONF_TILT_MIN),
-        config.get(CONF_TILT_MAX),
-        config.get(CONF_TILT_STATE_OPTIMISTIC),
-        config.get(CONF_TILT_INVERT_STATE),
-        config.get(CONF_SET_POSITION_TOPIC),
-        set_position_template,
-        config.get(CONF_UNIQUE_ID),
-        config.get(CONF_DEVICE),
-        discovery_hash
-    )])
+    async_add_entities([MqttCover(hass, config, discovery_hash)])
 
 
 class MqttCover(MqttAvailability, MqttDiscoveryUpdate, MqttEntityDeviceInfo,
                 CoverDevice):
     """Representation of a cover that can be controlled using MQTT."""
 
-    def __init__(self, name, state_topic, get_position_topic,
-                 command_topic, availability_topic,
-                 tilt_command_topic, tilt_status_topic, qos, retain,
-                 state_open, state_closed, position_open, position_closed,
-                 payload_open, payload_close, payload_stop, payload_available,
-                 payload_not_available, optimistic, value_template,
-                 tilt_open_position, tilt_closed_position, tilt_min, tilt_max,
-                 tilt_optimistic, tilt_invert, set_position_topic,
-                 set_position_template, unique_id: Optional[str],
-                 device_config: Optional[ConfigType], discovery_hash):
+    def __init__(self, hass, config, discovery_hash):
```

Comment:
Is it OK to pass `hass` to the constructor?

Reply:
@balloob Any comments on if this is OK? hass is passed to the constructor here since it's otherwise not set until async_added_to_hass() is called, and we need it when setting up the templates.

Reply:
@balloob Yes, possible, but maybe even cleaner to move to _subscribe_topics()?